cmd/util/cmd/find-trie-root/cmd.go

```diff
@@ -0,0 +1,232 @@
+	flagRootHash          string
+	flagFrom              int
+	flagTo                int
+	flagOutputDir         string
+)
+
+// find trie root hash from the wal files.
+// useful for state extraction and rolling back executed height.
+// for instance, when extracting state for a target height, it requires the wal files
+// has the trie root hash of the target block as the latest few records. If not the case,
+// then it is necessary to trim the wal files to the last record with the target trie root hash.
+// in order to do that, this command can be used to find the trie root hash in the wal files,
+// and copy the wal that contains the trie root hash to a new directory and trim it to
+// have the target trie root hash as the last record.
+// after that, the new wal file can be used to extract the state for the target height.
+var Cmd = &cobra.Command{
+	Use:   "find-trie-root",
+	Short: "find trie root hash from the wal files",
+	Run:   run,
+}
+
+func init() {
+	Cmd.Flags().StringVar(&flagExecutionStateDir, "execution-state-dir", "/var/flow/data/execution", "directory to the execution state")
+	_ = Cmd.MarkFlagRequired("execution-state-dir")
+
+	Cmd.Flags().StringVar(&flagRootHash, "root-hash", "",
+		"ledger root hash (hex-encoded, 64 characters)")
+	_ = Cmd.MarkFlagRequired("root-hash")
+
+	Cmd.Flags().IntVar(&flagFrom, "from", 0, "from segment")
+	Cmd.Flags().IntVar(&flagTo, "to", math.MaxInt32, "to segment")
+
+	Cmd.Flags().StringVar(&flagOutputDir, "output-dir", "", "output directory")
+}
+
+func run(*cobra.Command, []string) {
+	rootHash, err := parseInput(flagRootHash)
+	if err != nil {
+		log.Fatal().Err(err).Msg("cannot parse input")
+	}
+
+	segment, offset, err := searchRootHashInSegments(rootHash, flagExecutionStateDir, flagFrom, flagTo)
+	if err != nil {
+		log.Fatal().Err(err).Msg("cannot find root hash in segments")
+	}
+	log.Info().Msgf("found root hash in segment %d at offset %d", segment, offset)
+
+	if len(flagOutputDir) == 0 {
+		return
+	}
+
+	err = copyWAL(flagExecutionStateDir, flagOutputDir, segment, rootHash)
+	if err != nil {
+		log.Fatal().Err(err).Msg("cannot copy WAL")
+	}
+
+	log.Info().Msgf("copied WAL to %s", flagOutputDir)
+}
+
+func parseInput(rootHashStr string) (ledger.RootHash, error) {
+	rootHashBytes, err := hex.DecodeString(rootHashStr)
+	if err != nil {
+		return ledger.RootHash(hash.DummyHash), fmt.Errorf("cannot decode root hash: %w", err)
+	}
+	rootHash, err := ledger.ToRootHash(rootHashBytes)
+	if err != nil {
+		return ledger.RootHash(hash.DummyHash), fmt.Errorf("invalid root hash: %w", err)
+	}
+	return rootHash, nil
+}
+
+func searchRootHashInSegments(
+	expectedHash ledger.RootHash,
+	dir string,
+	wantFrom, wantTo int,
+) (int, int64, error) {
+	lg := zerolog.New(os.Stderr).With().Timestamp().Logger()
+	w, err := prometheusWAL.NewSize(lg, prometheus.DefaultRegisterer, dir, wal.SegmentSize, false)
+	if err != nil {
+		return 0, 0, fmt.Errorf("cannot create WAL: %w", err)
+	}
```

Comment:
Do we need to create `w` here?  We can use `dir` to replace `w.Dir()` as parameter `NewSegmentsRangeReader()`.